### PR TITLE
add BOXURLSessionCacheClient to persist data for background download/…

### PR DESCRIPTION
**Pull request by thuyBox**
…upload tasks to assist reconnection

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h [hunk 1]
```diff
@@ -442,6 +442,24 @@ extern NSString *const BOXAPIEnterpriseEventTypeCollaborationExpiration;
 extern NSString *const BOXAPIEnterpriseEventTypeItemSync;
 extern NSString *const BOXAPIEnterpriseEventTypeItemUnsync;
 
+//urlsessiontask cache dir, file prefix
+extern NSString *const BOXURLSessionTaskCacheDirName;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h]: Minor: would prefer Dir to expand to Directory for clarity. Also, proper caps and space after // for consistency.

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h [hunk 2]
```diff
@@ -77,6 +77,12 @@ typedef NS_ENUM(NSUInteger, BOXContentSDKStreamError) {
     BOXContentSDKStreamErrorReadFailed = 30001
 };
 
+typedef NS_ENUM(NSUInteger, BOXContentSDKCacheError) {
+    BOXContentSDKSCacheErrorArchiveFailed = 40000,
+    BOXContentSDKSCacheErrorFileCreateFailed = 40001,
+    BOXContentSDKSCacheErrorNoValidSessionTaskGivenUserIdAssociateId = 40002
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: Minor: The naming for the last one is a bit confusing. Is `BOXContentSDKCacheErrorNoValidSessionForUserId` inaccurate?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+//
+//  Created by Thuy Nguyen on 3/3/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@interface BOXURLSessionTaskCachedInfo : NSObject
+
+@property (nonatomic, strong, readwrite) NSString *backgroundSessionId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: Copy for strings.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+@property (nonatomic, assign, readwrite) NSUInteger sessionTaskId;
+@property (nonatomic, strong, readwrite) NSString *destinationFilePath;
+@property (nonatomic, strong, readwrite) NSData *resumeData;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:17]: What does `resumeData` mean?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+//
+
+#import "BOXURLSessionCacheClient.h"
+#import "BOXContentSDKConstants.h"
+#import "BOXLog.h"
+#import "BOXContentSDKErrors.h"
+
+@interface BOXURLBackgroundSessionIdAndSessionTaskId : NSObject
+
+@property (nonatomic, strong, readwrite) NSString *backgroundSessionId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Copy for strings.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+@property (nonatomic, assign, readwrite) NSUInteger sessionTaskId;
+
+@end
+
+@interface BOXURLSessionCacheClient()
+
+@property (nonatomic, strong, readwrite) NSString *cacheDir;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Copy for strings.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+@end
+
+@implementation BOXURLSessionCacheClient
+
+- (id)initWithCacheRootDir:(NSString *)cacheRootDir
+{
+    self = [super init];
+    if (self != nil) {
+        BOOL isDir = NO;
+        NSString *cacheDir = [cacheRootDir stringByAppendingPathComponent:BOXURLSessionTaskCacheDirName];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: We should either have a default and/or check if `cacheRootDir` is nil.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        if ([[NSFileManager defaultManager] fileExistsAtPath:cacheDir isDirectory:&isDir] == NO || isDir == NO) {
+            NSError *error;
+            [[NSFileManager defaultManager] createDirectoryAtPath:cacheDir withIntermediateDirectories:YES attributes:nil error:&error];
+            BOXAssert(error != nil, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Do we want to establish backup policy here in case we cannot create directory?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:data type:BOXURLSessionTaskCacheFileTypeResponse error:error];
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId taskError:(NSData *)taskError error:(NSError **)error
+{
+    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:taskError];
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:data type:BOXURLSessionTaskCacheFileTypeError error:error];
+}
+
+- (BOXURLSessionTaskCachedInfo *)getCacheForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Minor: let's just use `cacheInfoForUserId` method name, instead of Java's standard.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+{
+    BOXURLBackgroundSessionIdAndSessionTaskId *backgroundSessionIdAndSessionTaskId = [self backgroundSessionIdAndSessionTaskIdGivenUserId:userId associateId:associateId];
+
+    if (backgroundSessionIdAndSessionTaskId == nil) {
+        if (error != nil) {
+        (*error) = [[NSError alloc] initWithDomain:BOXURLSessionTaskCacheError code:BOXContentSDKSCacheErrorNoValidSessionTaskGivenUserIdAssociateId userInfo:nil];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Minor: why parenthesis around error?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+
+    NSString *backgroundSessionId = backgroundSessionIdAndSessionTaskId.backgroundSessionId;
+    NSUInteger sessionTaskId = backgroundSessionIdAndSessionTaskId.sessionTaskId;
+
+    NSString *dir = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+
+    BOXURLSessionTaskCachedInfo *cachedInfo = [[BOXURLSessionTaskCachedInfo alloc] init];
+
+    //get all files under sessions/$backgroundSessionId/$sessionTaskId
+    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:dir error:error];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Minor: filePaths instead of files.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+        }
+    }
+    return cachedInfo;
+}
+
+- (BOOL)deleteCachedInfoForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
+{
+    BOXURLBackgroundSessionIdAndSessionTaskId *backgroundSessionIdAndSessionTaskId = [self backgroundSessionIdAndSessionTaskIdGivenUserId:userId associateId:associateId];
+    if (backgroundSessionIdAndSessionTaskId == nil) {
+        return YES;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Do we want to return NO here or that if it isn't found, it's expected that the contract here will return YES?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+
+    if (success == YES) {
+        //clean up users/$userId/$associateId/*
+        dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
+        success = [self deleteDirectory:dir error:error];
+    }
+    return success;
+}
+
+#pragma mark - private helpers
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:228]: Minor: proper caps.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    if (outError != nil) {
+        (*outError) = error;
+    }
+    return success;
+}
+
+- (BOOL)cacheAndEncryptData:(NSData *)data atPath:(NSString *)path error:(NSError **)outError
+{
+    NSData *finalData = data;
+    if ([self.delegate respondsToSelector:@selector(encryptData:)]) {
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: If the delegate doesn't respond to the selector, can we really say that we encrypted the file? Might want to reflect an "attempt" abstraction in the method signature if this isn't always the case.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+- (BOOL)createDirForBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId error:(NSError **)error
+{
+    NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    return [self createDirectory:path error:error];
+}
+
+- (BOOL)createDirectory:(NSString *)path error:(NSError **)error
+{
+    BOOL isDir = NO;
+    BOOL success = YES;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:294]: We should default success to NO.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+    if ([[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDir] == NO || isDir == NO) {
+        success = [[NSFileManager defaultManager] createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:error];
+    }
+    return success;
+}
+
+- (BOOL)deleteDirectory:(NSString *)directoryPath error:(NSError **)error
+{
+    BOOL isDir = NO;
+    BOOL success = YES;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:304]: We should default success to NO.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+- (BOOL)createFile:(NSString *)path error:(NSError **)error
+{
+    BOOL success = [[NSFileManager defaultManager] createFileAtPath:path contents:nil attributes:nil];
+    if (success == NO && error != nil) {
+        (*error) = [[NSError alloc] initWithDomain:BOXContentSDKErrorDomain code:BOXContentSDKSCacheErrorFileCreateFailed userInfo:nil];
+    }
+    return success;
+}
+
+// path sessions/$backgroundSessionId/$sessionTaskId
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: MInor: can we properly caps/format these short method comments? e.g. `// Filepath Pattern: ...`

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+- (NSString *)dirPathOfSessionTaskWithBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId
+{
+    return [[[self.cacheDir stringByAppendingPathComponent:BOXURLSessionTaskCacheSessionsDirName] stringByAppendingPathComponent:backgroundSessionId] stringByAppendingPathComponent:[NSString stringWithFormat:@"%lu", (unsigned long)sessionTaskId]];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: stringByAppendingString is hard to parse. Let's use NSString stringWithFormat. Easier to read/change.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+// path users/$userId/$associateId/$backgroundSessionId-$sessionTaskId
+- (NSString *)dirPathOfUserSessionTaskGivenUserId:(NSString *)userId associateId:(NSString *)associateId
+{
+    return [[[self.cacheDir stringByAppendingPathComponent:BOXURLSessionTaskCacheUsersDirName] stringByAppendingPathComponent:userId] stringByAppendingPathComponent:associateId];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Same for this.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+    NSString *fileName = [self fileNameGivenBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    return [path stringByAppendingPathComponent:fileName];
+}
+
+- (NSString *)fileNameGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId
+{
+    return [NSString stringWithFormat:@"%@-%lu", backgroundSessionId, (unsigned long)sessionTaskId];
+}
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)parseBackgroundSessionIdAndSessionTaskIdFileName:(NSString *)name
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:348]: Minor: this method name is hard to read. Can we simplify this to something else?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+{
+    NSArray *arr = [name componentsSeparatedByString:@"-"];
+    BOXURLBackgroundSessionIdAndSessionTaskId *res = nil;
+    if (arr.count == 2) {
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:352]: Can we get an example of what a string might look like to parse? Also, let's find a different variable name for `arr`.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+    }
+    return res;
+}
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId
+{
+    NSString *dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
+    NSError *error;
+    BOXURLBackgroundSessionIdAndSessionTaskId *res = nil;
+    NSArray *arr = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:dir error:&error];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Arr -> filePaths

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+        res.sessionTaskId = [arr[1] unsignedIntegerValue];
+    }
+    return res;
+}
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId
+{
+    NSString *dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
+    NSError *error;
+    BOXURLBackgroundSessionIdAndSessionTaskId *res = nil;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: What's res? Let's use backgroundSessionIdAdSessionTaskId.

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.m [hunk 24]
```diff
@@ -421,3 +421,14 @@
 NSString *const BOXAPIEnterpriseEventTypeCollaborationExpiration = @"COLLABORATION_EXPIRATION";
 NSString *const BOXAPIEnterpriseEventTypeItemSync = @"ITEM_SYNC";
 NSString *const BOXAPIEnterpriseEventTypeItemUnsync = @"ITEM_UNSYNC";
+
+
+//urlsessiontask cache dir, file prefix
+NSString *const BOXURLSessionTaskCacheDirName = @"BOXURLSessionCache";
+NSString *const BOXURLSessionTaskCacheSessionsDirName = @"sessions";
+NSString *const BOXURLSessionTaskCacheUsersDirName = @"users";
+NSString *const BOXURLSessionTaskCacheDestinationFilePath = @"destinationFilePath";
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.m]: postfix the following values with 'Key'?

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h [hunk 25]
```diff
@@ -77,6 +77,12 @@ typedef NS_ENUM(NSUInteger, BOXContentSDKStreamError) {
     BOXContentSDKStreamErrorReadFailed = 30001
 };
 
+typedef NS_ENUM(NSUInteger, BOXContentSDKCacheError) {
+    BOXContentSDKSCacheErrorArchiveFailed = 40000,
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: overlapping with BOXContentSDKAppUserError.

Also, do we need one for 'SDKCacheErrorUnknown'?

Also, should the name be more specific to the networking/session/whatever cache for clarity?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+@property (nonatomic, strong, readwrite) NSURLResponse *response;
+@property (nonatomic, strong, readwrite) NSData *responseData;
+@property (nonatomic, strong, readwrite) NSError *error;
+
+@end
+
+@protocol BOXURLSessionCacheClientDelegate <NSObject>
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:24]: generic name, but this is specifically an encryption delegate, we should call that out in the name.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+
+// allow delegate to encrypt data before BOXURLSessionCacheClient persists it to disk
+- (NSData *)encryptData:(NSData *)data;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: These should be optional, no?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+ * - responseData:        store the response data (could contain error from server)
+ * - response:            store NSURLResponse
+ * - error:               store client-side NSError
+ *
+ * 2. To keep track of whose user and its associateId the session task belongs to, we save backgroundSessionId and sessionTaskId
+ * as a file name under sub-directory /users/$userId/$associateId with format $backgroundSessionId-$sessionTaskId
+ */
+@interface BOXURLSessionCacheClient : NSObject
+
+@property (nonatomic, weak, readwrite) id<BOXURLSessionCacheClientDelegate> delegate;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: Should have a comment explaining that it is OK to leave this unset and that it would just mean the networking cache information will not be encrypted.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * @return BOXURLSessionTaskCache   all cached data of the session task
+ */
+- (BOXURLSessionTaskCachedInfo *)getCacheForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
+
+/**
+ * Delete all cached data of the background session task associated with this userId and associateId
+ *
+ * @param userId            Id of user started the session task
+ * @param associateId       Id to associate with the session task
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: can you pass nil for this to remove all information for a particular user? If so, should that be explained here in the comments?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+ * @param error             error if fail to delete
+ *
+ * @return YES if succeeded, NO if failed
+ */
+- (BOOL)deleteCachedInfoForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:161]: Minor: Apple's naming convention involves capitalizing full acronyms (like URL), in this case ID. Also, for ID Apple tends to go with 'Identifier'

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+@implementation BOXURLSessionCacheClient
+
+- (id)initWithCacheRootDir:(NSString *)cacheRootDir
+{
+    self = [super init];
+    if (self != nil) {
+        BOOL isDir = NO;
+        NSString *cacheDir = [cacheRootDir stringByAppendingPathComponent:BOXURLSessionTaskCacheDirName];
+        if ([[NSFileManager defaultManager] fileExistsAtPath:cacheDir isDirectory:&isDir] == NO || isDir == NO) {
+            NSError *error;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Should initialize to nil, to form the standard/habit on the team - even though Apple improved this for objects, for primitives this matters.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 9 above)
Comment (boxtcarpel) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: +1 we should not be using 'get' in Obj-C generally speaking, it has an association with inserting the value into a variable passed in by reference. If you want to indicate that it's being loaded from disk, etc, you can use 'fetch...'

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 18 above)
Comment (boxtcarpel) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Please note this isn't appendingString, it's appendingPathComponent which is the right thing to use for correct path formatting. Using stringWithFormat with manual slashes etc is error-prone and not change-safe.

We can still break this up for better readability if we think it's too much as-is.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 26 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:24]: thinking we don't need to be specific for encryption only, there might be other use cases. Added "@optional"

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 27 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 28 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 29 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: as of now, don't think we have to support deleting all cache for a userId, so I'll make it explicit that associateId cannot be nil

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h (hunk 1 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h]: done

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h (hunk 25 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: done

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h (hunk 2 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 3 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 4 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:17]: I have comment explaining those properties, I can add a @see below note

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 31 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 5 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 6 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 7 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 8 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: we attempt to create dir whenever we want to cache data again if dir doesn't exist.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 9 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 10 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 11 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 12 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: added the check for error from backgroundSessionIdAndSessionTaskIdGivenUserId method, if there is an error, we return NO instead

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 14 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 15 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:294]: we don't want to since the dir might exists already

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 16 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:304]: we don't want to since the dir might already be deleted

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 17 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 21 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:352]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 22 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 23 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.m (hunk 24 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.m]: don't think we need to for this case

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 26 above)
Comment (boxtcarpel) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:24]: IMO, unless we know/really think we're going to use it for something else, lets be specific at this point, it will make this stuff easier to consume/understand

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 32]
```diff
@@ -0,0 +1,378 @@
...
+        res.backgroundSessionId = arr[0];
+        res.sessionTaskId = [arr[1] unsignedIntegerValue];
+    }
+    return res;
+}
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
+{
+    NSString *dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
+    NSError *err;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:362]: minor: initialize value

**Review by boxtcarpel — COMMENTED** [review]
+1

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 4 above)
Comment (irvingruan) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:17]: To me, it sounds confusing because "resume" is a verb. Maybe "resumableData"?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 15 above)
Comment (irvingruan) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:294]: If it already exists, it should return NO IMO. Technically, it did *not* create the directory this time, which is what the method says.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+@property (nonatomic, strong, readwrite) NSData *responseData;
+@property (nonatomic, strong, readwrite) NSError *error;
+
+@end
+
+@protocol BOXURLSessionCacheClientDelegate <NSObject>
+
+@optional
+// allow delegate to encrypt data before BOXURLSessionCacheClient persists it to disk
+- (NSData *)encryptData:(NSData *)data;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:28]: As with delegate patterns per Apple's recommendations, we should include the object in the delegate method, `- (NSData *)sessionCacheClient:(BOXURLSessionCacheClient *)client willEncryptData:(NSData *)data`.
